src/mergers/mergeAllOf.ts

```diff
@@ -40,7 +40,8 @@ function _mergeAllOf(fragment: SchemaFragment, path: string[], resolveRef: Walke
 
             if (Array.isArray(resolved.allOf)) {
               for (const member of resolved.allOf) {
-                if (typeof member.$ref === 'string' && schemaRefs.includes(member.$ref)) {
+                const index = schemaRefs.indexOf(member.$ref);
+                if (typeof member.$ref === 'string' && index !== -1 && index !== schemaRefs.lastIndexOf(member.$ref)) {
```

Comment:
I still don't quite like it, but at the same time, we'd need a more robust solution inside of json-schema-merge-allof itself to make come up with something considerably better.